scenario.py

```diff
@@ -189,10 +190,216 @@ def __init__(self, params, experiment_path):
+        labels = list(set(y_train))
+        random.seed(42)
+        random.shuffle(labels)
+
+        # Check coherence of the split option:
+        nb_diff_labels = len(labels)
+        sum_specific_clusters = sum([p.cluster_count for p in partners_with_specific_clusters])
+        if not partners_with_shared_clusters:
+            max_shared_clusters = 0
+        else:
+            max_shared_clusters = max([p.cluster_count for p in partners_with_shared_clusters])
+        assert sum_specific_clusters + max_shared_clusters  <= nb_diff_labels
+
+        # Stratify the dataset per labels
+        x_train_for_cluster, y_train_for_cluster, count_available_per_cluster = {}, {}, {}
+        for label in labels:
+            idx_in_full_trainset = np.where(y_train == label)
+            x_train_for_cluster[label] = x_train[idx_in_full_trainset]
+            y_train_for_cluster[label] = y_train[idx_in_full_trainset]
+            count_available_per_cluster[label] = len(y_train_for_cluster[label])
+
+        # Sort partners per nb of different clusters in descending order
+        partners_with_shared_clusters = sorted(
+            partners_with_shared_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+        partners_with_specific_clusters = sorted(
+            partners_with_specific_clusters,
+            key=operator.attrgetter("cluster_count"),
+            reverse=True,
+        )
+
+        # Initialize dict for storing which partners pick from which clusters
+        dict_specific, dict_shared = {}, {}
+
+        # First, partners getting data samples from clusters they don't share
+        position = 0
+        for p in partners_with_specific_clusters:
+            dict_specific[p.id] = {"clusters_list": labels[position:position + p.cluster_count]}
+            position += p.cluster_count
+
+        # Second, partners getting data samples from shared clusters
+        shared_clusters = labels[position:position + max_shared_clusters]
+        for p in partners_with_shared_clusters:
+            dict_shared[p.id] = {"clusters_list": random.sample(shared_clusters, k=p.cluster_count)}
+
+        # print(shared_clusters)  # DEBUG
+        # print(dict_specific)  # DEBUG
+        # print(dict_shared)  # DEBUG
+
+        # Compute the data amount limiting factor
+        resize_factor = 1
+
+        # For partners getting data samples from clusters they don't share
+        # ... compare the nb of available samples vs. the nb of samples initially configured
+        for p in partners_with_specific_clusters:
+            count_available = sum([count_available_per_cluster[cl] for cl in dict_specific[p.id]["clusters_list"]])
+            count_initialconfig = int(amounts_per_partner[p.id] * len(y_train))
+            dict_specific[p.id]["nb_available_samples"] = count_available
+            dict_specific[p.id]["nb_samples_configured"] = count_initialconfig
```

Comment:
We could use 2 fields in the Parner class for this, instead of a dict

Reply:
We could yes. But I am not too fan of this solution. Because we have a number of temporary variables (nb_available_samples, nb_samples_configured, nb_samples_configured_resized, nb_samples_configured_resized_per_cluster, final_nb_samples, final_nb_samples_p_cluster) which are of no use after this split. So we won't use the `Partner` object for those. And so using it for two of them doesn't seem to make much sense.

Reply:
So a `dict` seems to work well here, don't you think?